posts/2022-10-06-experimenting-with-voting-schemes.adoc

```diff
@@ -0,0 +1,400 @@
+
+The security of ElGamal is based on the hardness of solving a https://en.wikipedia.org/wiki/Discrete_logarithm[discrete logarithm], which is the operation of finding `x` given `h`, `g` and `p` such that `h = g^x^ mod p`. In other words, it is easy to calculate `h = g^x^` but much harder to recover `x = log~p~(h)` in modular arithmetic. Let us see how this can be used to form an encryption scheme.
+
+Setup:: Let the secret key `x` be a random number and let the public key be `h = g^x^ mod p`. `p` is a large prime number publicly known, and all the following operations are calculated modulo `p`. `g` is a generator of the https://en.wikipedia.org/wiki/Cyclic_group[cyclic group] of order `p`.
+
+Encryption:: Let the __ephemeral key__ `r` be a random number, and calculate the __shared secret__ `s = h^r^`. The ciphertext is the pair `(g^r^, m*s)`.
+
+Decryption:: The shared secret can be recovered from the first part of the ciphertext and the secret key: `s = (g^r^)^x^ = (g^x^)^r^ = h^r^`. Since the inverse of an element of the cyclic group can be https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm[calculated efficiently] we can recover the original message in the following way: `(m*s) * s^-1^ = m * (s*s^-1^) = m`.
+
+As mentioned before, two ciphertexts can be combined homomorphically to form an encryption of the product of the two messages. Indeed, the element-wise multiplication of the ciphertexts `(g^r~1~^, m~1~*h^r~1~^)` and `(g^r~2~^, m~2~*h^r~2~^)` yields `(g^r~1~+r~2~^, (m~1~*m~2~*h^r~1~+r~2~^)`, which is the encryption of `m~1~ * m~2~` with the ephemeral key `r~1~ + r~2~`.
+
+==== Example
+
+In the following example we examine the encryption, combination and decryption of Alice's and Bob's votes using ElGamal encryption. This illustrates how the ballots are encrypted in practice and should give a more concrete intuition of how the votes stay private.
+
+.Details of Alice's and Bob's encrypted ballots
+[%collapsible]
+====
+
+* Alice wants to vote for the first candidate and Bob for the second: `v~a~ = (1, 0, 0)` and `v~b~ = (0, 1, 0)`
+* let `q = 53`, `p = 2q + 1 = 107` and `g = 75`: `+g+` generates a subgroup of `Z/pZ` of order `q`
+* (note that we have not mentioned `q` earlier for simplicity, but it does not change the rest of this example)
+* let the key pair be `x = 18` and `h = g^x^ = 35 mod p`
+* the encrypted ballots can be calculated as follows: 
+** choose a random value, say `r = 2`
+** encryption of the first element of Alice's vote: `b~a1~ = (g^r^, h^r^ * g^m^) = (61, 48 * 75) = (61, 69)`
+** note that the vote is projected in `Z/pZ` before being encrypted with `g^m^`
+** similarly for the first element of Bob's vote with `r = 3`: `b~b1~ = (81, 75 * 1) = (81, 75)`
+* the encrypted sum for the first candidate is: `c~1~ = b~a1~ * b~b1~ = (61 * 81, 69 * 75) = (19, 39)`
+* and once decrypted: `s~1~ = 39 * (19^x^)^-1^ = 39 * 69^-1^ = 39 * 76 = 75 = g^1^`
+* the sum of the votes for the first candidate is indeed `1`
```

Comment:
I think that this section could benefit from some extra clarifications.

For example, what does "the vote is projected `Z/pZ`" mean. In the same sentence what is `m` in `g^m`?

In the sentence "and once decrypted: ..." what is done to get from `19^x^-1` to `69^-1`?

I did the calculation in the Scala REPL. I wonder if putting a small VSC/Metals worksheet in the project wouldn't be a good idea...

Reply:
good points
I added a link to a scastie snippet, this way there is no need to clone the repo